service/src/main/java/org/broadinstitute/listener/relay/inspectors/SamPermissionInspector.java

```diff
@@ -0,0 +1,69 @@
+package org.broadinstitute.listener.relay.inspectors;
+
+import com.microsoft.azure.relay.RelayedHttpListenerRequest;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Optional;
+import org.broadinstitute.listener.relay.inspectors.InspectorType.InspectorNameConstants;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.lang.NonNull;
+import org.springframework.stereotype.Component;
+
+@Component(InspectorNameConstants.SAM_CHECKER)
+public class SamPermissionInspector implements RequestInspector {
+  private final Logger logger = LoggerFactory.getLogger(SamPermissionInspector.class);
+  private final SamResourceClient samResourceClient;
+
+  public SamPermissionInspector(SamResourceClient samResourceClient) {
+    this.samResourceClient = samResourceClient;
+  }
+
+  @Override
+  public boolean inspectWebSocketUpgradeRequest(
+      @NonNull RelayedHttpListenerRequest relayedHttpListenerRequest) {
+    return checkPermission(relayedHttpListenerRequest.getHeaders());
+  }
+
+  @Override
+  public boolean inspectRelayedHttpRequest(
+      @NonNull RelayedHttpListenerRequest relayedHttpListenerRequest) {
+    return checkPermission(relayedHttpListenerRequest.getHeaders());
+  }
+
+  private boolean checkPermission(Map<String, String> headers) {
+
+    if (headers == null) {
+      logger.error("No auth headers found");
+      return false;
+    }
+
+    var cookieValue = headers.getOrDefault("cookie", headers.get("Cookie"));
+
+    var leoToken = getToken(cookieValue);
+
+    if (leoToken.isEmpty()) {
+      logger.error("No valid cookie found");
+      return false;
+    } else {
+      var token = leoToken.get();
+      return checkCachedPermission(token);
+    }
+  }
+
+  public boolean checkCachedPermission(String accessToken) {
+    var expiresAt = samResourceClient.checkWritePermission(accessToken);
+    return expiresAt.isAfter(Instant.now());
+  }
+
+  protected Optional<String> getToken(@NotNull String cookieValue) {
```

Comment:
I'm confused by the use of `@NotNull` in `@NotNull String cookieValue` as it looks like it could be `null`, if neither `cookie` or `Cookie` are in the http request header list. Not sure why the not-null checker, if there is one, doesn't produce a warning here.